### PR TITLE
Support for case insensitive column matching for ORC/Arrow/Parquet files

### PR DESCRIPTION
**Pull request by antonio2368**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add settings `input_format_orc_case_insensitive_column_matching`, `input_format_arrow_case_insensitive_column_matching`, and `input_format_parquet_case_insensitive_column_matching` which allows ClickHouse to use case insensitive matching of columns while reading data from ORC, Arrow or Parquet files.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Common/StringUtils/StringUtils.h [hunk 1]
```diff
@@ -240,6 +240,15 @@ inline bool equalsCaseInsensitive(char a, char b)
     return a == b || (isAlphaASCII(a) && alternateCaseIfAlphaASCII(a) == b);
 }
 
+inline bool equalsCaseInsensitive(const std::string_view a, const std::string_view b)
```
Comment (Avogar) [inline src/Common/StringUtils/StringUtils.h:243]: Consider using `boost::iequals` from #include `<boost/algorithm/string.hpp>`

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 2]
```diff
@@ -355,15 +366,13 @@ static ColumnWithTypeAndName readColumnFromArrowColumn(
         // ClickHouse writes Date as arrow UINT16 and DateTime as arrow UINT32,
         // so, read UINT16 as Date and UINT32 as DateTime to perform correct conversion
         // between Date and DateTime further.
-        case arrow::Type::UINT16:
-        {
+        case arrow::Type::UINT16: {
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:369]: Our code style assumes using
```
case ...
{
    ...
}
```
Please, remove such changes (I guess they were done by your IDE)

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 3]
```diff
@@ -1,39 +1,44 @@
 #include "ArrowColumnToCHColumn.h"
 
+#include <base/logger_useful.h>
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:3]: I guess you forgot to delete it after debugging.

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 4]
```diff
@@ -1,39 +1,44 @@
...
+#include <boost/algorithm/string/case_conv.hpp>
+#include "Common/StringUtils/StringUtils.h"
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:5]: Let's add these includes next to the others below

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 5]
```diff
@@ -508,28 +539,42 @@ Block ArrowColumnToCHColumn::arrowSchemaToCHHeader(const arrow::Schema & schema,
...
     {
         std::shared_ptr<arrow::ChunkedArray> arrow_column = table->GetColumnByName(column_name);
         if (!arrow_column)
             throw Exception(ErrorCodes::DUPLICATE_COLUMN, "Column '{}' is duplicated", column_name);
-        name_to_column_ptr[column_name] = arrow_column;
+
+        if (case_insensitive_matching)
+        {
+            boost::to_lower(column_name);
+        }
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:576]: Just fyi, our code style allows to write:
```
if (cond)
    one_line_code;
```
Sometimes it looks better then using braces

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 6]
```diff
@@ -484,19 +498,36 @@ static void checkStatus(const arrow::Status & status, const String & column_name
...
+            return true;
+
+        if (!ignore_case)
+            return nested_table_names.contains(field_name);
+
+        return std::find_if(
+                   nested_table_names.begin(),
+                   nested_table_names.end(),
+                   [&](const auto & nested_table_name) { return equalsCaseInsensitive(nested_table_name, field_name); })
+            != nested_table_names.end();
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:521]: I would suggest add flag `bool to_lower_case` in `Nested::getAllTableNames` to get the set of names in lower case and use just `nested_table_names.contains()` with `field_name` in lower case. Here and in the other similar places. But if you think current way is better, it's ok.

File: src/Processors/Formats/Impl/ORCBlockInputFormat.cpp [hunk 7]
```diff
@@ -165,19 +151,39 @@ void ORCBlockInputFormat::prepareReader()
         /// so we should recursively count the number of indices we need for this type.
         int indexes_count = countIndicesForType(schema->field(i)->type());
         const auto & name = schema->field(i)->name();
-        if (getPort().getHeader().has(name) || nested_table_names.contains(name))
+        const bool contains_column = std::invoke(
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ORCBlockInputFormat.cpp:154]: To be honest, don't see much reasons to use `std::invoke` here. If we add flag in `Nested::getAllTableNames` to get the set of names in lower case, we will be able to make this code much easier, like:
```c++
bool case_insensitive = format_settings.parquet.case_insensitive_column_matching;
if (getPort().getHeader().has(name, case_insensitive)
        || nested_table_names.contains(case_insensitive ? boost::to_lower_copy(name) : name))
{
    ...
}
```

**Comment by Avogar** [comment]
I see you added empty file `case_insensitive_column_matching.parquet.json`, I guess we can delete it.
And one small comment, this PR contains a lot of minor unrelated changes in code style (IDE could makes some changes for us), sometimes it's a bit irritating, because it becomes difficult to read the real changes

File: src/Processors/Formats/Impl/ORCBlockInputFormat.cpp [hunk 8]
```diff
@@ -165,19 +151,39 @@ void ORCBlockInputFormat::prepareReader()
...
+                return std::find_if(
+                           nested_table_names.begin(),
+                           nested_table_names.end(),
+                           [&](const auto & nested_table_name) { return equalsCaseInsensitive(nested_table_name, name); })
+                    != nested_table_names.end();
+            });
+
+        if (contains_column)
         {
+            column_names.push_back(name);
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply src/Processors/Formats/Impl/ORCBlockInputFormat.cpp]: So I reverted the #35145 which deleted the unused `column_names`, will remove it again.

File: src/Processors/Formats/Impl/ORCBlockInputFormat.cpp (hunk 7 above)
Comment (antonio2368) [reply src/Processors/Formats/Impl/ORCBlockInputFormat.cpp]: `std::invoke` was used to eliminate a lot of scopes with if checks, and it's much cleaner than `continue` IMO. 
But with your suggestion for `Nested::getAllTableNames` it's not needed.

**Review by Avogar — APPROVED** [review]
Let's wait for the tests and merge it if everything will be ok

**Comment by antonio2368** [comment]
@Mergifyio update

**Comment by antonio2368** [comment]
@Mergifyio update

**Pull request merged**